File: src/ios/MSKBedfontPlugin.swift
//
//  MSKBedfontPlugin.swift
//
//  Created by Shishir Shetty on 2023-10-02.
//
import Foundation
import SmokerlyzerSDK



var smokerlyzerBluetooth = SmokerlyzerBluetooth()



@objc(MSKBedfontPlugin)
class MSKBedfontPlugin: CDVPlugin, ConnectionObserver {
    
     let BEDFONT_EVENT_RECOVERYCHANGE = "OnRecoveryChangeEvent"
    
     let BEDFONT_EVENT_SCANSTATECHANGE = "OnScanStateChangeEvent"
    
     let BEDFONT_EVENT_CONNECTIONCHANGE = "OnConnectionChangeEvent"
    
     let BEDFONT_EVENT_CONNECTINGCHANGE = "OnCOnnectingChangeEvent"
    
     let BEDFONT_EVENT_BUTTONNAMECHANGE = "OnButtonNameChangeEvent"
    
     let BEDFONT_EVENT_CONNECT_RESULT = "OnConnectResultEvent"
    
     let BEDFONT_EVENT_SCAN_RESULT = "OnScanResultEvent"
    
     let BEDFONT_EVENT_DEVICE_USAGE = "onDeviceUsageEvent"
    
     let BEDFONT_EVENT_DEVICE_FIRMWARE = "onDeviceFirmwareEvent"
    
     let BEDFONT_EVENT_DEVICE_SERIALNUMBER = "onDeviceSerialNumberEvent"
    
    
    
     let SUCCESS = "SUCCESS"
     let SUCCESS_NEEDS_RECOVERY = "SUCCESS_NEEDS_RECOVERY"
     let ZEROING = "ZEROING"
     let ERROR_FAILED_TO_FINALIZE = "ERROR_FAILED_TO_FINALIZE"
     let ERROR_FAILED_TO_CONNECT = "ERROR_FAILED_TO_CONNECT"
     let ERROR_SCAN_FAILED = "ERROR_SCAN_FAILED"
    
     let DEVICE_NOT_CONNECTED = "DEVICE_NOT_CONNECTED"
    
    
    var STATE_DISCONNECT = true
    var STATE_SCAN_AND_CONNECT = false
    
    var initialized = false
    
    var sensor: PeripheralIdentifier?
    var isConnectedText: String = "Disconnected"
    
    var isEnabled: Bool = false
    
    //var myCommandDelegate: CDVCommandDelegate?
    

    override func pluginInitialize() {
        super.pluginInitialize()

        // Initialize myCommandDelegate with self.commandDelegate during plugin initialization
        //myCommandDelegate = self.commandDelegate
    }
    
    @objc(initialize:)func initialize(command : CDVInvokedUrlCommand) {
        
        if(!initialized) {
            smokerlyzerBluetooth = SmokerlyzerBluetooth()
            smokerlyzerBluetooth.register(connectionObserver: self)
            
            initialized = true
            let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
            self.commandDelegate.send(result, callbackId: command.callbackId)
        } else {
            let result = CDVPluginResult.init(status: CDVCommandStatus_OK, messageAs: "Already Initialized!")
            self.commandDelegate.send(result, callbackId: command.callbackId)
        }
    }
    
    @objc(dispose:)func dispose(command : CDVInvokedUrlCommand) {
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(getStatus:)func getStatus(command : CDVInvokedUrlCommand) {
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.getIsConnected { result in
                if(result) {
                    let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
                    self.commandDelegate.send(result, callbackId: command.callbackId)
                } else {
                    let errorMessage = "Device is not connected"
                    let result = CDVPluginResult(status: CDVCommandStatus_ERROR, messageAs: errorMessage)
                    self.commandDelegate.send(result, callbackId: command.callbackId)
                }
            }
        }
        
    }
    
    @objc(performScanning:)func performScanning(command : CDVInvokedUrlCommand) {
        
        self.perform_Scanning(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(performDisconnection:)func performDisconnection(command : CDVInvokedUrlCommand) {
        
        self.perform_Disconnection(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(testNorecovery:)func testNorecovery(command : CDVInvokedUrlCommand) {
        
        self.perform_TestNorecovery(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(getDeviceFirmware:)func getDeviceFirmware(command : CDVInvokedUrlCommand) {
        
        self.get_DeviceFirmware(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(getDeviceSerialNumber:)func getDeviceSerialNumber(command : CDVInvokedUrlCommand) {
        
        self.get_DeviceSerialNumber(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    @objc(getDeviceBattery:)func getDeviceBattery(command : CDVInvokedUrlCommand) {
        
        self.get_DeviceBattery(commandDelegate: self.commandDelegate)
        let result = CDVPluginResult.init(status: CDVCommandStatus_OK)
        self.commandDelegate.send(result, callbackId: command.callbackId)
    }
    
    
    func bluetoothAvailable(_ available: Bool) {
        //log(message: "Bluetooth available: \(available)")
    }

    func connected(to peripheral: PeripheralIdentifier) {
        //log(message: "[connection observer] detected connect event")
        //sensor = peripheral
        //isConnectedText = "Connected"
        self.sendStateChangeEvents(commandDelegate: self.commandDelegate, boolval: true)
    }

    func disconnected(from peripheral: PeripheralIdentifier) {
        //log(message: "[connection observer] detected disconnect event")
        //isConnectedText = "Disconnected"
        self.sendStateChangeEvents(commandDelegate: self.commandDelegate, boolval: false)
    }
    
    
    
    func perform_Scanning(commandDelegate: CDVCommandDelegate) {
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.scanAndConnect(connected: {update in
                    switch update {
                            case .success(let peripheralId):
                                //self.log(message: "Successfully connected to " + peripheralId.name)
                                //self.isConnectedText = "Connected"
                                self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "SUCCESS", logMessage: "Finalized connection. Device is READY")
                            case .successNeedsRecovery(let peripheralId):
                                //self.log(message: "Connected to " + peripheralId.name + ". Recovery is needed before starting breath test.")
                                //self.isConnectedText = "Connected"
                                self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "SUCCESS_NEEDS_RECOVERY", logMessage: "Finalized connection. Recovery function needs to be run on the sensor")
                            case .zeroing:
                                //self.log(message: "Device is zeroing, please wait")
                                self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "ZEROING", logMessage: "Zeroing the sensor, please wait...")
                            case .failure(let error):
                                if let error = error as? BluejayError {
                                    switch error{
                                    case .scanFailed:
                                        //self.log(message: "Scanning failed to pick up a device")
                                        self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "ERROR_SCAN_FAILED", logMessage: "Failed to find device")
                                    case .connectionTimedOut:
                                        //self.log(message: "Failed to connect to device")
                                        self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "ERROR_FAILED_TO_CONNECT", logMessage: "Failed to connect to device")
                                    default:
                                        //self.log(message: "Error: " + error.localizedDescription)
                                        self.sendScanningEvents(commandDelegate: commandDelegate, connectResult: "ERROR_SCAN_FAILED", logMessage: "Failed to find device")
                                    }
                                }
                        //self.isConnectedText = "Disconnected"
                    }
                                                })
                }
    }
   
    
    func perform_Disconnection(commandDelegate: CDVCommandDelegate) {
        self.sendStateChangeEvents(commandDelegate: commandDelegate, boolval: false)
        self.isEnabled = false
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.disconnect(force: true)
        }
    }
    
    func perform_TestNorecovery(commandDelegate: CDVCommandDelegate) {
        if smokerlyzerBluetooth != nil {
            
            smokerlyzerBluetooth.startBreathTestNoRecovery {result in
                switch result {
                case .success(let ppmResult):
                    self.sendScanningResults(commandDelegate: commandDelegate, statusName: "PPM value is " + ppmResult.latest.description, ppm: ppmResult.latest, isSuccessful: true)
                case .failure(let error):
                    self.sendScanningResults(commandDelegate: commandDelegate, statusName: "Error: " + error.localizedDescription, ppm: -1, isSuccessful: false)
                }
            }
        }
        
    }
    func get_DeviceFirmware(commandDelegate: CDVCommandDelegate) {
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.getFirmwareVersion(callback: { result in
                switch result {
                case .success(let firmware):
                    //self.log(message: "Firmware version: " + firmware.version)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceFirmwareEvent", deviceParam: firmware.version)
                case .failure(let error):
                    //self.log(message: "Error: " + error.localizedDescription)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceFirmwareEvent", deviceParam: "Error: " + error.localizedDescription)
                }
            })
        }
    }
    
    func get_DeviceSerialNumber(commandDelegate: CDVCommandDelegate) {
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.getSerialNumber(callback: { result in
                switch result {
                case .success(let serial):
                    //self.log(message: "Serial number: " + serial.serial)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceSerialNumberEvent", deviceParam: serial.serial)
                case .failure(let error):
                    //self.log(message: "Error: " + error.localizedDescription)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceSerialNumberEvent", deviceParam: "Error: " + error.localizedDescription)
                }
            })
        }
    }
    
    func get_DeviceBattery(commandDelegate: CDVCommandDelegate) {
        if smokerlyzerBluetooth != nil {
            smokerlyzerBluetooth.getBatteryReading(callback: { result in
                switch result {
                case .success(let battery):
                    //self.log(message: "Serial number: " + serial.serial)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceBatteryEvent", deviceParam: String(battery.volts))
                case .failure(let error):
                    //self.log(message: "Error: " + error.localizedDescription)
                    self.sendDeviceDetails(commandDelegate: commandDelegate, eventName: "onDeviceBatteryEvent", deviceParam: "Error: " + error.localizedDescription)
                }
            })
        }
    }
    
    func sendStateChangeEvents(commandDelegate: CDVCommandDelegate, boolval: Bool) {
        let eventData: [String: Any] = [
            "changeState": boolval
        ]
        self.fireEvent(commandDelegate: commandDelegate, eventName: "OnButtonNameChangeEvent", eventData: eventData )
    }
    
    func sendScanningEvents(commandDelegate: CDVCommandDelegate,connectResult: String, logMessage: String) {
        let eventData: [String: Any] = [
            "connectResult": connectResult,
            "logMessage": logMessage
        ]
        self.fireEvent(commandDelegate: commandDelegate, eventName: "OnConnectResultEvent", eventData: eventData)
    }
    
    func sendScanningResults(commandDelegate: CDVCommandDelegate, statusName: String, ppm: Int, isSuccessful: Bool) {
        let eventData: [String: Any] = [
            "statusName": statusName,
            "ppm": ppm,
            "isSuccessful": isSuccessful
        ]
        self.fireEvent(commandDelegate: commandDelegate, eventName: "OnScanResultEvent", eventData: eventData)
    }
    
    func sendDeviceDetails(commandDelegate: CDVCommandDelegate, eventName: String, deviceParam: String) {
        let eventData: [String: Any] = [
            "deviceDetail": deviceParam
        ]
        
        self.fireEvent(commandDelegate: commandDelegate, eventName: eventName, eventData: eventData)
    }
    
    func fireEvent(commandDelegate: CDVCommandDelegate, eventName: String, eventData: [String: Any]) {
        
        do {
            // Convert the dictionary to JSON data
            let jsonData = try JSONSerialization.data(withJSONObject: eventData, options: [])

            // Print the JSON data as a string (for demonstration purposes)
            if let jsonString = String(data: jsonData, encoding: .utf8) {
                
                let jsCode = "cordova.fireDocumentEvent('" + eventName + "'," + jsonString + ")"
                commandDelegate.evalJs(jsCode)
            }
        } catch {

        }
        
    }
}


//extension MSKBedfontPlugin: ConnectionObserver {
//    func bluetoothAvailable(_ available: Bool) {
//        //log(message: "Bluetooth available: \(available)")
//    }
//
//    func connected(to peripheral: PeripheralIdentifier) {
//        //log(message: "[connection observer] detected connect event")
//        //sensor = peripheral
//        //isConnectedText = "Connected"
//        self.sendStateChangeEvents(commandDelegate: self.commandDelegate, boolval: true)
//    }
//
//    func disconnected(from peripheral: PeripheralIdentifier) {
//        //log(message: "[connection observer] detected disconnect event")
//        //isConnectedText = "Disconnected"
//        self.sendStateChangeEvents(commandDelegate: self.commandDelegate, boolval: false)
//    }
//}

File: src/android/MSKBedfontPlugin.java
package org.mskcc.bedfont;

import org.apache.cordova.CordovaPlugin;
import org.apache.cordova.CallbackContext;
import org.apache.cordova.CordovaWebView;
import org.apache.cordova.CordovaInterface;
import org.apache.cordova.PluginResult;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Map;
import java.util.Objects;

import android.app.Application;
import android.util.Log;

import com.bedfont.icosdk.ble.v2.*;

import io.reactivex.disposables.Disposable;

public class MSKBedfontPlugin extends CordovaPlugin {

    /***LIST OF ALL EVENTS****/
    private static final String BEDFONT_EVENT_RECOVERYCHANGE = "OnRecoveryChangeEvent";

    private static final String BEDFONT_EVENT_SCANSTATECHANGE = "OnScanStateChangeEvent";

    private static final String BEDFONT_EVENT_CONNECTIONCHANGE = "OnConnectionChangeEvent";

    private static final String BEDFONT_EVENT_CONNECTINGCHANGE = "OnCOnnectingChangeEvent";

    private static final String BEDFONT_EVENT_BUTTONNAMECHANGE = "OnButtonNameChangeEvent";

    private static final String BEDFONT_EVENT_CONNECT_RESULT = "OnConnectResultEvent";

    private static final String BEDFONT_EVENT_SCAN_RESULT = "OnScanResultEvent";

    private static final String BEDFONT_EVENT_DEVICE_USAGE = "onDeviceUsageEvent";

    private static final String BEDFONT_EVENT_DEVICE_FIRMWARE = "onDeviceFirmwareEvent";

    private static final String BEDFONT_EVENT_DEVICE_SERIALNUMBER = "onDeviceSerialNumberEvent";

    private static final String BEDFONT_EVENT_DEVICE_BATTERY = "onDeviceBatteryEvent";



    private static final String SUCCESS = "SUCCESS";
    private static final String SUCCESS_NEEDS_RECOVERY = "SUCCESS_NEEDS_RECOVERY";
    private static final String ZEROING = "ZEROING";
    private static final String ERROR_FAILED_TO_FINALIZE = "ERROR_FAILED_TO_FINALIZE";
    private static final String ERROR_FAILED_TO_CONNECT = "ERROR_FAILED_TO_CONNECT";
    private static final String ERROR_SCAN_FAILED = "ERROR_SCAN_FAILED";

    private static final String DEVICE_NOT_CONNECTED = "DEVICE_NOT_CONNECTED";

    private static final boolean STATE_DISCONNECT = true;
    private static final boolean STATE_SCAN_AND_CONNECT = false;

    private CordovaWebView webView;
    private CallbackContext callback;
    private Boolean initialized = false;

    private SmokerlyzerBluetoothLeManager smokerlyzerBluetoothLeManager;
    private Disposable connectionSubscription;
    private Disposable scanningSubscription;
    private Disposable onBackToSubscription;
    private Disposable connectingSubscription;
    private Disposable recoverySubscription;

    @Override
    public void initialize(CordovaInterface cordova, CordovaWebView webView) {
        super.initialize(cordova, webView);

        this.webView = webView;
    }

    @Override
    public boolean execute(String action, JSONArray args, CallbackContext callbackContext) throws JSONException {
        callback = callbackContext;
        PluginResult result;

        switch (action) {

            case "initialize":
                if(!initialized) {
                    initSDK();
                    initialized=true;
                    result = new PluginResult(PluginResult.Status.OK);
                }
                else {
                    result = new PluginResult(PluginResult.Status.OK, "Already Initialized!");
                }

                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;

            case "dispose":
                // Cleanup subscriptions
                if(scanningSubscription!=null)
                    scanningSubscription.dispose();
                if(connectionSubscription!=null)
                    connectionSubscription.dispose();
                if(connectingSubscription!=null)
                    connectingSubscription.dispose();
                if(recoverySubscription!=null)
                    recoverySubscription.dispose();

                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;

            case "performScanning":
                performScanning();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;

            case "performDisconnection":
                perfromDisconnection();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;

            case "testNorecovery":
                performTestNoRecovery();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;

            case "getDeviceUsage":
                getDeviceUsage();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;
            case "getDeviceFirmware":
                getDeviceFirmware();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;
            case "getDeviceSerialNumber":
                getDeviceSerialNumber();
                result = new PluginResult(PluginResult.Status.OK);
                result.setKeepCallback(false);
                callback.sendPluginResult(result);
                return true;
            case "getDeviceBattery":
                 getDeviceBattery();
                 result = new PluginResult(PluginResult.Status.OK);
                 result.setKeepCallback(false);
                 callback.sendPluginResult(result);
                 return true;

        }

        return false;

    }


    private void initSDK() {
        smokerlyzerBluetoothLeManager = SmokerlyzerBluetoothLeManager.build(this.cordova.getContext());
        scanningSubscription = smokerlyzerBluetoothLeManager
                .getScanTracker()
                .getSubscription()
                .subscribe(this::scanStateChange);
        // Subscribe to listen for a disconnect
        connectionSubscription = smokerlyzerBluetoothLeManager
                .getConnectionTracker()
                .getSubscription()
                .subscribe(this::connectionChange);
        connectingSubscription = smokerlyzerBluetoothLeManager
                .getConnectingTracker()
                .getSubscription()
                .subscribe(this::connectingChange);
        recoverySubscription = smokerlyzerBluetoothLeManager
                .getRecoveryListener()
                .getSubscription()
                .subscribe(this::recoveryChange);
    }

    private void recoveryChange(Boolean aBoolean) {
        sendStateChangeEvents(BEDFONT_EVENT_RECOVERYCHANGE, aBoolean.booleanValue());
    }

    private void connectingChange(Boolean aBoolean) {
        sendStateChangeEvents(BEDFONT_EVENT_CONNECTINGCHANGE, aBoolean.booleanValue());
    }

    private void connectionChange(Boolean aBoolean) {
        sendStateChangeEvents(BEDFONT_EVENT_CONNECTIONCHANGE, aBoolean.booleanValue());
        if(aBoolean) {
            connected();
        } else {
            disconnected();
        }
    }


    private void scanStateChange(Boolean aBoolean) {
        sendStateChangeEvents(BEDFONT_EVENT_SCANSTATECHANGE, aBoolean.booleanValue());
    }




    private void connected() {
        sendStateChangeEvents(BEDFONT_EVENT_BUTTONNAMECHANGE, STATE_DISCONNECT);
    }

    private void disconnected() {
        sendStateChangeEvents(BEDFONT_EVENT_BUTTONNAMECHANGE, STATE_SCAN_AND_CONNECT);
    }


    private void performScanning() {
        if(smokerlyzerBluetoothLeManager!=null) {
            smokerlyzerBluetoothLeManager.scanAndConnect(new String[]{"iCOquit"}, connectResult -> {
                switch(connectResult) {
                    case SUCCESS:
                        sendScanningEvents(SUCCESS, "Finalized connection. Device is READY");
                        break;
                    case SUCCESS_NEEDS_RECOVERY:
                        sendScanningEvents(SUCCESS_NEEDS_RECOVERY,"Finalized connection. Recovery function needs to be run on the sensor");
                        break;
                    case ZEROING:
                        sendScanningEvents(ZEROING,"Zeroing the sensor, please wait...");
                        break;
                    case ERROR_FAILED_TO_FINALIZE:
                        sendScanningEvents(ERROR_FAILED_TO_FINALIZE,"Connection process failed to finalize.");
                        break;
                    case ERROR_FAILED_TO_CONNECT:
                        sendScanningEvents(ERROR_FAILED_TO_CONNECT,"Failed to connect to device");
                        break;
                    case ERROR_SCAN_FAILED:
                        sendScanningEvents(ERROR_SCAN_FAILED,"Failed to find device");
                        break;
                }
            });
        }
    }

    private void perfromDisconnection() {

        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.disconnect();
        disconnected();
    }


    private void performTestNoRecovery() {
        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.getIsConnected((isConnected) -> {
                if(isConnected) {
                    smokerlyzerBluetoothLeManager.startBreathTestNoRecovery(this::onBreathTestComplete);
                }
                else {
                    sendScanningEvents(DEVICE_NOT_CONNECTED,"Device is not connected");
                }
            });
    }


    public void onBreathTestComplete(boolean isSuccessful, int ppm, SmokerlyzerBluetoothLeManager.StatusCodeConstants status) {

        if(isSuccessful && status == SmokerlyzerBluetoothLeManager.StatusCodeConstants.SUCCESS){
            sendScanningResults(status.name(), ppm, isSuccessful);
        } else if (status == SmokerlyzerBluetoothLeManager.StatusCodeConstants.ERROR_ZEROING_REQUIRED) {
            sendScanningResults(status.name(), -1, isSuccessful);
        }
        else if (status == SmokerlyzerBluetoothLeManager.StatusCodeConstants.ERROR_DEVICE_NOT_READY) {
            sendScanningResults(status.name(), -1, isSuccessful);
        }
        else{
            sendScanningResults(status.name(), -1, isSuccessful);
        }

    }

    private void getDeviceUsage() {
        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.getIsConnected((r2) -> {
                if (r2) {
                    smokerlyzerBluetoothLeManager.getUsage((r) -> {
                        sendDeviceDetails(BEDFONT_EVENT_DEVICE_USAGE, ""+r);
                    });
                } else{
                    sendDeviceDetails(BEDFONT_EVENT_DEVICE_USAGE, "Device Not Connected");
                }
            });
    }

    private void getDeviceFirmware() {
        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.getIsConnected((r2) -> {
                if (r2) {
                    smokerlyzerBluetoothLeManager.getFirmwareVersion((r) -> {
                        sendDeviceDetails(BEDFONT_EVENT_DEVICE_FIRMWARE, ""+r);
                    });
                } else{
                    sendDeviceDetails(BEDFONT_EVENT_DEVICE_FIRMWARE, "Device Not Connected");
                }
            });
    }

    private void getDeviceSerialNumber() {
        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.getIsConnected((r2) -> {
                if (r2) {
                    smokerlyzerBluetoothLeManager.getSerialNumber((r) -> {
                        sendDeviceDetails(BEDFONT_EVENT_DEVICE_SERIALNUMBER, ""+r);
                    });
                } else{
                    sendDeviceDetails(BEDFONT_EVENT_DEVICE_SERIALNUMBER, "Device Not Connected");
                }
            });
    }

    private void getDeviceBattery() {
        if(smokerlyzerBluetoothLeManager!=null)
            smokerlyzerBluetoothLeManager.getIsConnected((r2) -> {
                if (r2) {
                    smokerlyzerBluetoothLeManager.getBatteryReading((r) -> {
                        sendDeviceDetails(BEDFONT_EVENT_DEVICE_BATTERY, ""+r);
                    });
                } else{
                    sendDeviceDetails(BEDFONT_EVENT_DEVICE_BATTERY, "Device Not Connected");
                }
            });
    }



    public void sendStateChangeEvents(String eventName, boolean boolVal) {

        JSONObject params = new JSONObject();
        try {
            params.put("changeState", boolVal);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        fireEvent(eventName, params);
    }

    public void sendScanningEvents(String connectResult, String logMessage) {

        JSONObject params = new JSONObject();
        try {
            params.put("connectResult", connectResult);
            params.put("logMessage", logMessage);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        fireEvent(BEDFONT_EVENT_CONNECT_RESULT, params);
    }

    public void sendScanningResults(String statusName, int ppm, boolean isSuccessful) {

        JSONObject params = new JSONObject();
        try {
            params.put("statusName", statusName);
            params.put("ppm", ppm);
            params.put("isSuccessful", isSuccessful);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        fireEvent(BEDFONT_EVENT_SCAN_RESULT, params);
    }

    public void sendDeviceDetails(String eventName, String deviceParam) {

        JSONObject params = new JSONObject();
        try {
            params.put("deviceDetail", deviceParam);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        fireEvent(eventName, params);
    }

    public void fireEvent(String eventName, JSONObject params) {
        showToast(params.toString());
        String js = "cordova.fireDocumentEvent('" + eventName + "', " + params.toString() + ");";
        webView.getView().post(() -> webView.loadUrl("javascript:" + js));
    }

    private void showToast(String message) {
//        final android.widget.Toast toast = android.widget.Toast.makeText(cordova.getActivity().getWindow().getContext(), message, android.widget.Toast.LENGTH_LONG);
//        toast.show();
        Log.d("Bedfont", message);
    }
}
